ci/dev/watch.ts

```diff
@@ -1,139 +1,204 @@
-import * as cp from "child_process"
+import { spawn, fork, ChildProcess } from "child_process"
+import del from "del"
+import { promises as fs } from "fs"
 import * as path from "path"
-import { onLine } from "../../src/node/util"
-
-async function main(): Promise<void> {
-  try {
-    const watcher = new Watcher()
-    await watcher.watch()
-  } catch (error: any) {
-    console.error(error.message)
-    process.exit(1)
-  }
+import { CompilationStats, onLine, OnLineCallback, VSCodeCompileStatus } from "../../src/node/util"
+
+interface DevelopmentCompilers {
+  [key: string]: ChildProcess | undefined
+  vscode: ChildProcess
+  vscodeWebExtensions: ChildProcess
+  codeServer: ChildProcess
+  plugins: ChildProcess | undefined
 }
 
 class Watcher {
-  private readonly rootPath = path.resolve(__dirname, "../..")
-  private readonly vscodeSourcePath = path.join(this.rootPath, "vendor/modules/code-oss-dev")
+  private rootPath = path.resolve(process.cwd())
+  private readonly paths = {
+    /** Path to uncompiled VS Code source. */
+    vscodeDir: path.join(this.rootPath, "vendor", "modules", "code-oss-dev"),
+    compilationStatsFile: path.join(this.rootPath, "out", "watcher.json"),
+    pluginDir: process.env.PLUGIN_DIR,
+  }
+
+  //#region Web Server
 
-  private static log(message: string, skipNewline = false): void {
-    process.stdout.write(message)
-    if (!skipNewline) {
-      process.stdout.write("\n")
+  /** Development web server. */
+  private webServer: ChildProcess | undefined
+
+  private reloadWebServer = (): void => {
+    if (this.webServer) {
+      this.webServer.kill()
     }
+
+    this.webServer = fork(path.join(this.rootPath, "out/node/entry.js"), process.argv.slice(2))
+    const { pid } = this.webServer
+
+    this.webServer.on("exit", () => console.log("[Code Server]", `Web process ${pid} exited`))
+
+    console.log("\n[Code Server]", `Spawned web server process ${pid}`)
   }
 
-  public async watch(): Promise<void> {
-    let server: cp.ChildProcess | undefined
-    const restartServer = (): void => {
-      if (server) {
-        server.kill()
-      }
-      const s = cp.fork(path.join(this.rootPath, "out/node/entry.js"), process.argv.slice(2))
-      console.log(`[server] spawned process ${s.pid}`)
-      s.on("exit", () => console.log(`[server] process ${s.pid} exited`))
-      server = s
-    }
+  //#endregion
 
-    const vscode = cp.spawn("yarn", ["watch"], { cwd: this.vscodeSourcePath })
+  //#region Compilers
```

Comment:
I like this pattern you've started introducing 👏 Does it have a name? Comment regions?

Reply:
I wish I could take credit for it but it’s built into [VS Code’s region collapsing](https://dev.to/rfornal/vs-code-collapse-and-expand-4710).